File: apps/mobile/android/app/src/main/java/com/diveforge/ble/PeregrineBleManager.kt
package com.diveforge.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.ParcelUuid
import com.diveforge.ble.protocol.*
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.ByteArrayOutputStream
import java.util.UUID

/**
 * Android BLE transport for Shearwater dive computers.
 *
 * Uses BluetoothGatt callbacks with Kotlin coroutines for async coordination.
 * Implements the same flow as the iOS PeregrineClient:
 *   scan -> connect -> discover services -> subscribe notifications ->
 *   transfer (wrap -> SLIP -> fragment -> write; await -> stripHeader -> SLIP decode -> unwrap)
 *
 * Key differences from iOS (CoreBluetooth):
 *   - Uses BluetoothLeScanner for scanning
 *   - Uses BluetoothGatt callbacks (on binder thread)
 *   - Must request MTU explicitly
 *   - Uses Channel<ByteArray> for frame awaiting
 *   - Uses newSingleThreadContext("BLE") for serialized operations
 */
@SuppressLint("MissingPermission")
class PeregrineBleManager(private val context: Context) {

    companion object {
        // Shearwater custom GATT service/characteristic UUIDs
        private val UART_SERVICE_UUID = UUID.fromString("fe25c237-0ece-443c-b0aa-e02033e7029d")
        private val UART_TX_CHAR_UUID = UUID.fromString("27b7570b-359e-45a3-91bb-cf7e70049bd2")
        private val UART_RX_CHAR_UUID = UUID.fromString("27b7570b-359e-45a3-91bb-cf7e70049bd2")
        private val CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

        private const val SCAN_TIMEOUT_MS = 15_000L
        private const val TRANSFER_TIMEOUT_MS = 10_000L
        private const val MAX_BLOCK_RETRIES = 2
        private const val DEFAULT_CHUNK_SIZE = 75  // Cap at 75 even if MTU allows more
    }

    // ---- State ----

    private var bluetoothGatt: BluetoothGatt? = null
    private var txCharacteristic: BluetoothGattCharacteristic? = null
    private var negotiatedMtu: Int = 23  // Android default
    private var chunkSize: Int = Peregrine.BLE_CHUNK_SIZE

    private val bleContext = newSingleThreadContext("BLE")
    private val transferMutex = Mutex()
    private val frameChannel = Channel<ByteArray>(Channel.UNLIMITED)
    private val slipDecoder = SlipCodec.Decoder()

    private var scanner: BluetoothLeScanner? = null
    private var scanCallback: ScanCallback? = null
    private var connectionContinuation: CancellableContinuation<Unit>? = null
    private var mtuContinuation: CancellableContinuation<Int>? = null
    private var writeContinuation: CancellableContinuation<Unit>? = null

    data class DiscoveredDevice(
        val name: String?,
        val address: String,
        val rssi: Int
    )

    // ---- Scanning ----

    suspend fun startScan(
        serviceUuid: String,
        onDiscovered: (DiscoveredDevice) -> Unit
    ) {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        val adapter = bluetoothManager.adapter
            ?: throw PeregrineProtocolException.NotConnected()

        scanner = adapter.bluetoothLeScanner
            ?: throw PeregrineProtocolException.NotConnected()

        val filter = ScanFilter.Builder()
            .setServiceUuid(ParcelUuid(UUID.fromString(serviceUuid)))
            .build()

        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        scanCallback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                onDiscovered(DiscoveredDevice(
                    name = result.device.name,
                    address = result.device.address,
                    rssi = result.rssi
                ))
            }
        }

        scanner?.startScan(listOf(filter), settings, scanCallback!!)

        // Auto-stop after timeout
        withContext(bleContext) {
            delay(SCAN_TIMEOUT_MS)
            stopScan()
        }
    }

    fun stopScan() {
        scanCallback?.let { cb ->
            scanner?.stopScan(cb)
        }
        scanCallback = null
        scanner = null
    }

    // ---- Connection ----

    suspend fun connect(address: String) {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        val adapter = bluetoothManager.adapter
            ?: throw PeregrineProtocolException.NotConnected()

        val device = adapter.getRemoteDevice(address)

        suspendCancellableCoroutine<Unit> { cont ->
            connectionContinuation = cont
            bluetoothGatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
            cont.invokeOnCancellation {
                bluetoothGatt?.disconnect()
                bluetoothGatt?.close()
                bluetoothGatt = null
            }
        }

        // Request larger MTU
        val mtu = suspendCancellableCoroutine<Int> { cont ->
            mtuContinuation = cont
            bluetoothGatt?.requestMtu(185) // Request 185 to get ~180 usable
        }
        negotiatedMtu = mtu
        // Chunk size = MTU - 5 (3 ATT overhead + 2 BLE mini-header), capped at 75
        chunkSize = minOf(mtu - 5, DEFAULT_CHUNK_SIZE)
    }

    fun disconnect() {
        bluetoothGatt?.disconnect()
        bluetoothGatt?.close()
        bluetoothGatt = null
        txCharacteristic = null
    }

    fun isConnected(): Boolean {
        return bluetoothGatt != null && txCharacteristic != null
    }

    // ---- GATT Callback ----

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    connectionContinuation?.cancel(PeregrineProtocolException.NotConnected())
                    connectionContinuation = null
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                connectionContinuation?.cancel(
                    PeregrineProtocolException.UnexpectedResponse("Service discovery failed: $status")
                )
                return
            }

            val service = gatt.getService(UART_SERVICE_UUID)
            if (service == null) {
                connectionContinuation?.cancel(
                    PeregrineProtocolException.UnexpectedResponse("UART service not found")
                )
                return
            }

            txCharacteristic = service.getCharacteristic(UART_TX_CHAR_UUID)
            val rxCharacteristic = service.getCharacteristic(UART_RX_CHAR_UUID)

            if (txCharacteristic == null || rxCharacteristic == null) {
                connectionContinuation?.cancel(
                    PeregrineProtocolException.UnexpectedResponse("UART characteristics not found")
                )
                return
            }

            // Enable notifications on RX characteristic
            gatt.setCharacteristicNotification(rxCharacteristic, true)
            val descriptor = rxCharacteristic.getDescriptor(CCCD_UUID)
            if (descriptor != null) {
                descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                gatt.writeDescriptor(descriptor)
            } else {
                // No CCCD, resolve immediately
                connectionContinuation?.resumeWith(Result.success(Unit))
                connectionContinuation = null
            }
        }

        override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            if (descriptor.uuid == CCCD_UUID) {
                if (status == BluetoothGatt.GATT_SUCCESS) {
                    connectionContinuation?.resumeWith(Result.success(Unit))
                } else {
                    connectionContinuation?.cancel(
                        PeregrineProtocolException.UnexpectedResponse("CCCD write failed: $status")
                    )
                }
                connectionContinuation = null
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            mtuContinuation?.resumeWith(Result.success(mtu))
            mtuContinuation = null
        }

        override fun onCharacteristicWrite(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                writeContinuation?.resumeWith(Result.success(Unit))
            } else {
                writeContinuation?.cancel(
                    PeregrineProtocolException.UnexpectedResponse("Write failed: $status")
                )
            }
            writeContinuation = null
        }

        @Deprecated("Deprecated in API 33+, but needed for backward compat")
        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            val value = characteristic.value ?: return
            // Strip BLE mini-header and feed to SLIP decoder
            try {
                val payload = BleFramer.stripHeader(value)
                val frames = slipDecoder.feed(payload)
                for (frame in frames) {
                    frameChannel.trySend(frame)
                }
            } catch (e: Exception) {
                // Tolerate framing errors on individual notifications
            }
        }
    }

    // ---- Low-level transport ----

    /**
     * Write a single BLE chunk (with mini-header already applied).
     */
    private suspend fun writeChunk(chunk: ByteArray) {
        val gatt = bluetoothGatt ?: throw PeregrineProtocolException.NotConnected()
        val char = txCharacteristic ?: throw PeregrineProtocolException.NotConnected()

        suspendCancellableCoroutine<Unit> { cont ->
            writeContinuation = cont
            char.value = chunk
            char.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            if (!gatt.writeCharacteristic(char)) {
                writeContinuation = null
                cont.cancel(PeregrineProtocolException.UnexpectedResponse("writeCharacteristic returned false"))
            }
        }
    }

    /**
     * Fragment and write a SLIP-encoded payload as BLE chunks.
     */
    private suspend fun sendFrames(slipData: ByteArray) {
        val chunks = BleFramer.fragment(slipData, chunkSize)
        for (chunk in chunks) {
            writeChunk(chunk)
        }
    }

    /**
     * Await a complete SLIP frame from the device with timeout.
     */
    private suspend fun awaitFrame(timeoutMs: Long = TRANSFER_TIMEOUT_MS): ByteArray {
        return withTimeout(timeoutMs) {
            frameChannel.receive()
        }
    }

    // ---- Protocol-level transport ----

    /**
     * Full transfer cycle: wrap -> SLIP encode -> fragment -> write; await -> unwrap.
     * Returns the inner response payload.
     */
    suspend fun transfer(payload: ByteArray): ByteArray = transferMutex.withLock {
        val wrapped = WrapCodec.wrapRequest(payload)
        val slipEncoded = SlipCodec.encode(wrapped)
        sendFrames(slipEncoded)
        val response = awaitFrame()
        return WrapCodec.unwrapResponse(response)
    }

    /**
     * RDBI convenience: send RDBI request, parse response.
     */
    suspend fun rdbi(id: Int): ByteArray {
        val payload = RDBI.request(id)
        val response = transfer(payload)
        return RDBI.parse(response, id)
    }

    // ---- High-level operations ----

    /**
     * List dives: read serial/firmware/hardware/logupload, then download manifest.
     * Returns parsed manifest records.
     */
    suspend fun listDives(onProgress: ((String) -> Unit)? = null): List<ManifestRecord> {
        onProgress?.invoke("Reading device info...")

        // Read device identifiers
        val serial = rdbi(Peregrine.ID_SERIAL)
        val firmware = rdbi(Peregrine.ID_FIRMWARE)
        val hardware = rdbi(Peregrine.ID_HARDWARE)
        val logupload = rdbi(Peregrine.ID_LOGUPLOAD)

        onProgress?.invoke("Reading manifest...")

        // Determine base address from logupload response
        val baseAddr = LogbookFormat.baseAddress(logupload)

        // Download manifest via block download (no compression for manifest)
        val manifestData = blockDownload(
            address = Peregrine.MANIFEST_ADDR,
            size = Peregrine.MANIFEST_SIZE,
            compression = false,
            onProgress = { pct -> onProgress?.invoke("Manifest: ${pct}%") }
        )

        return Manifest.parse(manifestData)
    }

    /**
     * Download a single dive by manifest index (0-based into the ManifestRecord list).
     * Returns decompressed dive data.
     */
    suspend fun downloadDive(
        record: ManifestRecord,
        baseAddr: Long,
        onProgress: ((Int) -> Unit)? = null
    ): ByteArray {
        val diveAddress = baseAddr + record.address
        val compressed = blockDownload(
            address = diveAddress,
            size = Peregrine.DIVE_SIZE,
            compression = true,
            onProgress = { pct -> onProgress?.invoke(pct) }
        )
        return Decompress.full(compressed)
    }

    /**
     * Block download with retry and progress reporting.
     * Implements shearwater_common.c block download flow.
     */
    private suspend fun blockDownload(
        address: Long,
        size: Long,
        compression: Boolean,
        onProgress: ((Int) -> Unit)? = null
    ): ByteArray {
        // Send init request
        val initPayload = BlockDownload.initRequest(address, size, compression)
        val initResponse = transfer(initPayload)
        val blockSize = BlockDownload.parseInitResponse(initResponse)

        val output = ByteArrayOutputStream()
        var blockNum = 0
        var totalExpected = size
        var totalReceived = 0L

        while (true) {
            var attempt = 0
            var blockData: ByteArray? = null

            // Retry loop with exponential backoff
            while (attempt <= MAX_BLOCK_RETRIES) {
                try {
                    val blockPayload = BlockDownload.blockRequest(blockNum)
                    val blockResponse = transfer(blockPayload)
                    blockData = BlockDownload.parseBlockResponse(blockResponse, blockNum)
                    break
                } catch (e: PeregrineProtocolException) {
                    attempt++
                    if (attempt > MAX_BLOCK_RETRIES) throw e
                    // Exponential backoff: 100ms, 200ms
                    delay(100L * attempt)
                }
            }

            val data = blockData ?: throw PeregrineProtocolException.Timeout()

            // Check for end-of-transfer (empty block or short block)
            if (data.isEmpty()) break

            output.write(data)
            totalReceived += data.size
            blockNum = (blockNum + 1) and 0xFF

            // Report progress
            if (totalExpected > 0) {
                val pct = ((totalReceived * 100) / totalExpected).toInt().coerceAtMost(100)
                onProgress?.invoke(pct)
            }

            // If we got less than blockSize, this is the last block
            if (data.size < blockSize) break
        }

        // Send quit
        val quitResponse = transfer(BlockDownload.quitRequest)
        BlockDownload.parseQuitResponse(quitResponse)

        return output.toByteArray()
    }

    /**
     * Clean up resources.
     */
    fun destroy() {
        stopScan()
        disconnect()
        frameChannel.close()
        bleContext.close()
    }
}

File: apps/mobile/ios/DiveComputer/PeregrineBLEManager.swift
// PeregrineBLEManager.swift
// DiveForge — Production BLE layer for Shearwater Peregrine dive computers.
//
// Adapted from spike/0c-ble-protocol/swift-sources/PeregrineClient.swift.
// Removes SwiftUI/Combine dependencies, logging, auto-connect behavior.
// Adds callback-based event delivery, thread-safety via serial DispatchQueue,
// and connect-by-identifier for React Native bridge integration.

import Foundation
import CoreBluetooth

// MARK: - PeregrineBLEManager

final class PeregrineBLEManager: NSObject {

    // MARK: - Constants

    static let serviceUUID = CBUUID(string: "FE25C237-0ECE-443C-B0AA-E02033E7029D")
    static let sppCharacteristicUUID = CBUUID(string: "27B7570B-359E-45A3-91BB-CF7E70049BD2")

    // MARK: - Callbacks

    /// Called when a peripheral is discovered during scan.
    /// Parameters: (identifier UUID string, name or nil, RSSI)
    var onDiscovered: ((_ identifier: String, _ name: String?, _ rssi: Int) -> Void)?

    /// Called with progress updates during downloads.
    /// Parameters: (description string, bytes downloaded so far)
    var onProgress: ((_ label: String, _ bytesDownloaded: Int) -> Void)?

    /// Called when the peripheral disconnects (including unexpected disconnects).
    /// Parameter: optional error description
    var onDisconnected: ((_ error: String?) -> Void)?

    // MARK: - Public state

    /// Whether the manager is connected and ready to send commands.
    private(set) var isReady: Bool = false

    /// Firmware version string captured from RDBI during listDives().
    private(set) var firmwareVersion: String?

    // MARK: - Internals (CoreBluetooth)

    private var central: CBCentralManager!
    private var peripheral: CBPeripheral?
    private var sppCharacteristic: CBCharacteristic?
    private var pendingScan = false
    private var scanServiceUUID: CBUUID?

    /// Serial queue for thread-safety. All state mutations and CB delegate work happen here.
    private let queue = DispatchQueue(label: "com.diveforge.peregrine-ble", qos: .userInitiated)

    /// Discovered peripherals cache (for connect-by-identifier).
    private var discoveredPeripherals: [String: CBPeripheral] = [:]

    // MARK: - Internals (protocol layer)

    private let slipDecoder = SLIP.Decoder()

    /// Continuation waiting for the next decoded SLIP frame.
    private var pendingFrameContinuation: CheckedContinuation<Data, Error>?

    /// Queue of frames already decoded but not yet awaited.
    private var pendingFrames: [Data] = []

    /// Per-request timeout (ms). libdc uses 3000 ms (shearwater_common.c:62).
    private let readTimeoutMs: UInt64 = 5000

    // MARK: - Connect completion

    private var connectContinuation: CheckedContinuation<Void, Error>?

    // MARK: - Lifecycle

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: queue, options: [
            CBCentralManagerOptionShowPowerAlertKey: true,
        ])
    }

    // MARK: - Public actions (Layer 2)

    /// Start scanning for Peregrine peripherals. Discovered peripherals are reported via onDiscovered.
    func startScan(serviceUuid: String? = nil) {
        let uuid = serviceUuid.flatMap { CBUUID(string: $0) } ?? Self.serviceUUID
        scanServiceUUID = uuid
        queue.async { [weak self] in
            guard let self = self else { return }
            guard self.central.state == .poweredOn else {
                self.pendingScan = true
                return
            }
            self.discoveredPeripherals.removeAll()
            self.central.scanForPeripherals(withServices: [uuid], options: [
                CBCentralManagerScanOptionAllowDuplicatesKey: false,
            ])
        }
    }

    /// Stop scanning.
    func stopScan() {
        queue.async { [weak self] in
            self?.central.stopScan()
        }
    }

    /// Connect to a specific peripheral by its UUID identifier string.
    /// Resolves when the GATT characteristic is discovered and notifications are enabled (isReady = true).
    func connect(identifier: String) async throws {
        try await withCheckedThrowingContinuation { (cc: CheckedContinuation<Void, Error>) in
            queue.async { [weak self] in
                guard let self = self else {
                    cc.resume(throwing: PeregrineProtocolError.notConnected)
                    return
                }
                guard let targetUUID = UUID(uuidString: identifier) else {
                    cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("Invalid UUID: \(identifier)"))
                    return
                }

                // Check if we discovered it during scan.
                if let p = self.discoveredPeripherals[identifier] {
                    self.connectContinuation = cc
                    self.peripheral = p
                    p.delegate = self
                    self.central.stopScan()
                    self.central.connect(p, options: nil)
                } else {
                    // Try to retrieve by known identifier.
                    let known = self.central.retrievePeripherals(withIdentifiers: [targetUUID])
                    if let p = known.first {
                        self.connectContinuation = cc
                        self.peripheral = p
                        p.delegate = self
                        self.central.stopScan()
                        self.central.connect(p, options: nil)
                    } else {
                        cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("Peripheral \(identifier) not found"))
                    }
                }
            }
        }
    }

    /// Disconnect from the current peripheral.
    func disconnect() {
        queue.async { [weak self] in
            guard let self = self else { return }
            if let p = self.peripheral, p.state != .disconnected {
                self.central.cancelPeripheralConnection(p)
            }
            self.cleanupConnection(error: nil)
        }
    }

    // MARK: - Public actions (Layer 3)

    /// Discover dives on the device. Reads serial/firmware/hardware/logupload, then the
    /// manifest. Returns array of (index, address, fingerprintHex) tuples.
    func listDives() async throws -> [(index: Int, address: UInt32, fingerprintHex: String)] {
        guard isReady else { throw PeregrineProtocolError.notConnected }

        // Per shearwater_petrel.c:160-220, libdc reads serial -> firmware -> hardware -> logupload
        let _ = try await rdbi(id: Peregrine.ID_SERIAL, expectedLen: 8)

        let firmware = try await rdbi(id: Peregrine.ID_FIRMWARE)
        firmwareVersion = String(data: firmware, encoding: .ascii)

        let _ = try await rdbi(id: Peregrine.ID_HARDWARE, expectedLen: 2)

        let logupload = try await rdbi(id: Peregrine.ID_LOGUPLOAD, expectedLen: 9)
        let baseAddr = try LogbookFormat.baseAddress(fromLogUploadResponse: logupload)

        // Download the manifest (uncompressed).
        let manifestBlob = try await downloadBlob(
            address: Peregrine.MANIFEST_ADDR,
            size: Peregrine.MANIFEST_SIZE,
            compression: false
        )
        let records = Manifest.parse(manifestBlob)

        // Cache for downloadDive.
        self._baseAddr = baseAddr
        self._records = records

        return records.map { rec in
            (index: rec.index,
             address: rec.address,
             fingerprintHex: rec.fingerprint.map { String(format: "%02x", $0) }.joined())
        }
    }

    /// Download the dive at the given 1-based index. Returns raw decompressed bytes.
    func downloadDive(at index: Int) async throws -> Data {
        guard isReady else { throw PeregrineProtocolError.notConnected }
        guard let baseAddr = _baseAddr else {
            throw PeregrineProtocolError.unexpectedResponse("call listDives() first")
        }
        guard let rec = _records.first(where: { $0.index == index }) else {
            throw PeregrineProtocolError.unexpectedResponse("dive index \(index) not in manifest")
        }

        onProgress?("Starting dive \(index)...", 0)

        // Compressed dive-body download.
        let compressedBlob = try await downloadBlob(
            address: baseAddr + rec.address,
            size: Peregrine.DIVE_SIZE,
            compression: true,
            progressLabel: "dive \(index)"
        )

        // Decompress: LRE + XOR.
        let decompressed = try Decompress.full(compressedBlob)

        onProgress?("Done. \(decompressed.count) bytes.", decompressed.count)

        return decompressed
    }

    // Cache populated by listDives(), used by downloadDive(at:).
    private var _baseAddr: UInt32?
    private var _records: [ManifestRecord] = []

    // MARK: - Layer 3 internals (transfer + block download)

    /// One round-trip: wrap -> SLIP-encode -> BLE-fragment -> write; then await one logical
    /// SLIP frame, BLE-strip per chunk, SLIP-decode (already done), unwrap.
    /// Mirrors `shearwater_common_transfer` shearwater_common.c:328-388.
    private func transfer(_ payload: Data) async throws -> Data {
        // 1) Send.
        let req = Wrap.wrapRequest(payload)
        let slip = SLIP.encode(req)
        try await sendBLEFrames(slip)

        // 2) Await one full SLIP frame.
        let frame = try await awaitFrame(timeoutMs: readTimeoutMs)

        // 3) Unwrap.
        return try Wrap.unwrapResponse(frame)
    }

    /// RDBI helper. If `expectedLen` is non-nil, asserts the returned data has that exact length.
    private func rdbi(id: UInt16, expectedLen: Int? = nil) async throws -> Data {
        let resp = try await transfer(RDBI.request(id: id))
        let inner = try RDBI.parse(response: resp, id: id)
        if let n = expectedLen, inner.count != n {
            throw PeregrineProtocolError.unexpectedResponse(
                "RDBI id=\(String(format:"%04x",id)) expected \(n) got \(inner.count)"
            )
        }
        return inner
    }

    /// Block download driver — mirrors `shearwater_common_download` shearwater_common.c:391-519.
    private func downloadBlob(address: UInt32, size: UInt32, compression: Bool,
                              progressLabel: String? = nil) async throws -> Data {
        // Init.
        let initReq = BlockDownload.initRequest(address: address, size: size, compression: compression)
        let initRsp = try await transfer(initReq)
        let _ = try BlockDownload.parseInitResponse(initRsp)

        // Loop blocks.
        var raw = Data()
        var blockNum: UInt8 = 1
        var nbytes: UInt32 = 0
        var done = false

        while nbytes < size && !done {
            let req = BlockDownload.blockRequest(blockNum: blockNum)
            let rsp = try await transfer(req)
            let payload = try BlockDownload.parseBlockResponse(rsp, expectedBlock: blockNum)
            raw.append(payload)
            nbytes &+= UInt32(payload.count)
            blockNum = blockNum &+ 1

            if compression {
                // Cheap done-check: try LRE-decode to detect end-of-stream marker.
                if (raw.count * 8) % 9 == 0 {
                    let (_, isFinal) = try Decompress.lre(raw)
                    if isFinal { done = true }
                }
            }

            if let label = progressLabel {
                onProgress?(label, raw.count)
            }
        }

        // Quit.
        let quitRsp = try await transfer(BlockDownload.quitRequest)
        try BlockDownload.parseQuitResponse(quitRsp)

        return raw
    }

    // MARK: - BLE-frame I/O

    /// Fragment a single SLIP-encoded logical frame into BLE chunks and write each chunk.
    private func sendBLEFrames(_ slipFrame: Data) async throws {
        guard let p = peripheral, let ch = sppCharacteristic else {
            throw PeregrineProtocolError.notConnected
        }
        let chunkSize = Peregrine.BLE_CHUNK_SIZE

        // Prefer writeWithoutResponse for speed.
        let writeType: CBCharacteristicWriteType =
            ch.properties.contains(.writeWithoutResponse) ? .withoutResponse : .withResponse

        let chunks = BLEFramer.fragment(slipFrame, chunkSize: chunkSize)
        for chunk in chunks {
            if writeType == .withoutResponse {
                while !p.canSendWriteWithoutResponse {
                    try await Task.sleep(nanoseconds: 2_000_000) // 2 ms
                }
            }
            p.writeValue(chunk, for: ch, type: writeType)
        }
    }

    /// Drain incoming notification chunks: strip 2-byte mini-header, feed SLIP decoder,
    /// dispatch any complete logical frames.
    private func handleIncoming(_ chunk: Data) {
        let stripped: Data
        do {
            stripped = try BLEFramer.stripHeader(chunk)
        } catch {
            return
        }
        let frames = slipDecoder.feed(stripped)
        for frame in frames {
            if let cc = pendingFrameContinuation {
                pendingFrameContinuation = nil
                cc.resume(returning: frame)
            } else {
                pendingFrames.append(frame)
            }
        }
    }

    /// Wait for one complete logical SLIP frame, with timeout.
    private func awaitFrame(timeoutMs: UInt64) async throws -> Data {
        // If we have a queued frame, return immediately.
        if !pendingFrames.isEmpty {
            return pendingFrames.removeFirst()
        }
        return try await withThrowingTaskGroup(of: Data.self) { group in
            group.addTask { [weak self] in
                try await withCheckedThrowingContinuation { (cc: CheckedContinuation<Data, Error>) in
                    self?.queue.async {
                        guard let self = self else {
                            cc.resume(throwing: PeregrineProtocolError.notConnected)
                            return
                        }
                        if !self.pendingFrames.isEmpty {
                            cc.resume(returning: self.pendingFrames.removeFirst())
                        } else {
                            self.pendingFrameContinuation = cc
                        }
                    }
                }
            }
            group.addTask {
                try await Task.sleep(nanoseconds: timeoutMs * 1_000_000)
                throw PeregrineProtocolError.timeout
            }
            let first = try await group.next()!
            group.cancelAll()
            // Clear any dangling continuation if the timeout won the race.
            self.queue.async { [weak self] in
                if let cc = self?.pendingFrameContinuation {
                    self?.pendingFrameContinuation = nil
                    cc.resume(throwing: PeregrineProtocolError.timeout)
                }
            }
            return first
        }
    }

    // MARK: - Internal helpers

    private func cleanupConnection(error: String?) {
        peripheral = nil
        sppCharacteristic = nil
        isReady = false
        _baseAddr = nil
        _records.removeAll()
        firmwareVersion = nil
        // Fail any in-flight frame wait.
        if let cc = pendingFrameContinuation {
            pendingFrameContinuation = nil
            cc.resume(throwing: PeregrineProtocolError.notConnected)
        }
        pendingFrames.removeAll()
        // Fail connect continuation if pending.
        if let cc = connectContinuation {
            connectContinuation = nil
            cc.resume(throwing: PeregrineProtocolError.notConnected)
        }
        onDisconnected?(error)
    }
}

// MARK: - CBCentralManagerDelegate

extension PeregrineBLEManager: CBCentralManagerDelegate {

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state == .poweredOn, pendingScan {
            pendingScan = false
            let uuid = scanServiceUUID ?? Self.serviceUUID
            central.scanForPeripherals(withServices: [uuid], options: [
                CBCentralManagerScanOptionAllowDuplicatesKey: false,
            ])
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData: [String : Any],
                        rssi RSSI: NSNumber) {
        let id = peripheral.identifier.uuidString
        discoveredPeripherals[id] = peripheral
        onDiscovered?(id, peripheral.name, RSSI.intValue)
    }

    func centralManager(_ central: CBCentralManager,
                        didConnect peripheral: CBPeripheral) {
        peripheral.discoverServices([Self.serviceUUID])
    }

    func centralManager(_ central: CBCentralManager,
                        didFailToConnect peripheral: CBPeripheral,
                        error: Error?) {
        let msg = error?.localizedDescription ?? "unknown error"
        if let cc = connectContinuation {
            connectContinuation = nil
            cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("Connect failed: \(msg)"))
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDisconnectPeripheral peripheral: CBPeripheral,
                        error: Error?) {
        cleanupConnection(error: error?.localizedDescription)
    }
}

// MARK: - CBPeripheralDelegate

extension PeregrineBLEManager: CBPeripheralDelegate {

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let e = error {
            if let cc = connectContinuation {
                connectContinuation = nil
                cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("discoverServices: \(e.localizedDescription)"))
            }
            return
        }
        for svc in peripheral.services ?? [] {
            if svc.uuid == Self.serviceUUID {
                peripheral.discoverCharacteristics([Self.sppCharacteristicUUID], for: svc)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didDiscoverCharacteristicsFor service: CBService,
                    error: Error?) {
        if let e = error {
            if let cc = connectContinuation {
                connectContinuation = nil
                cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("discoverCharacteristics: \(e.localizedDescription)"))
            }
            return
        }
        for ch in service.characteristics ?? [] {
            if ch.uuid == Self.sppCharacteristicUUID {
                sppCharacteristic = ch
                peripheral.setNotifyValue(true, for: ch)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didUpdateNotificationStateFor characteristic: CBCharacteristic,
                    error: Error?) {
        if let e = error {
            if let cc = connectContinuation {
                connectContinuation = nil
                cc.resume(throwing: PeregrineProtocolError.unexpectedResponse("setNotifyValue: \(e.localizedDescription)"))
            }
            return
        }
        if characteristic.uuid == Self.sppCharacteristicUUID && characteristic.isNotifying {
            isReady = true
            if let cc = connectContinuation {
                connectContinuation = nil
                cc.resume()
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didUpdateValueFor characteristic: CBCharacteristic,
                    error: Error?) {
        guard error == nil, let value = characteristic.value else { return }
        handleIncoming(value)
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didWriteValueFor characteristic: CBCharacteristic,
                    error: Error?) {
        // No-op for production; write errors surface as timeouts on awaitFrame.
    }
}
